File: groups.php
<?php
header('Content-Type: application/json');
require_once __DIR__ . '/utils.php';

$method = $_SERVER['REQUEST_METHOD'];
$payload = null;
if ($method !== 'GET') {
  $raw = file_get_contents('php://input');
  $payload = json_decode($raw, true);
  if (!is_array($payload)) $payload = [];
}

$actorRole = $method === 'GET' ? ($_GET['role'] ?? 'user') : ($payload['actorRole'] ?? 'user');
if ($actorRole !== 'admin') {
  http_response_code(403);
  echo json_encode(["success" => false, "message" => "Operazione permessa solo agli amministratori"]);
  exit;
}

$groups = read_json_file(GROUPS_FILE);
if (!is_array($groups)) $groups = [];

function normalize_groups($groups) {
  $normalized = [];
  foreach ($groups as $group) {
    if (!is_string($group)) continue;
    $trimmed = trim($group);
    if ($trimmed === '') continue;
    if (!in_array($trimmed, $normalized, true)) {
      $normalized[] = $trimmed;
    }
  }
  return $normalized;
}

function persist_groups($groups) {
  if (!write_json_file(GROUPS_FILE, array_values($groups))) {
    http_response_code(500);
    echo json_encode(["success" => false, "message" => "Impossibile salvare i gruppi"]);
    exit;
  }
}

function persist_users_list($users) {
  if (!write_json_file(USERS_FILE, $users)) {
    http_response_code(500);
    echo json_encode(["success" => false, "message" => "Impossibile aggiornare gli utenti"]);
    exit;
  }
}

switch ($method) {
  case 'GET':
    echo json_encode(["success" => true, "groups" => array_values($groups)], JSON_UNESCAPED_UNICODE);
    break;

  case 'POST':
    $name = trim((string)($payload['name'] ?? ''));
    if ($name === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Il nome del gruppo è obbligatorio"]);
      exit;
    }
    if (in_array($name, $groups, true)) {
      http_response_code(409);
      echo json_encode(["success" => false, "message" => "Esiste già un gruppo con questo nome"]);
      exit;
    }
    $groups[] = $name;
    persist_groups($groups);
    echo json_encode(["success" => true]);
    break;

  case 'PUT':
    $oldName = trim((string)($payload['oldName'] ?? ''));
    $newName = trim((string)($payload['newName'] ?? ''));
    if ($oldName === '' || $newName === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Specificare i nomi del gruppo da modificare"]);
      exit;
    }
    if (!in_array($oldName, $groups, true)) {
      http_response_code(404);
      echo json_encode(["success" => false, "message" => "Gruppo non trovato"]);
      exit;
    }
    if ($oldName === $newName) {
      echo json_encode(["success" => true]);
      exit;
    }
    if (in_array($newName, $groups, true)) {
      http_response_code(409);
      echo json_encode(["success" => false, "message" => "Esiste già un gruppo con il nuovo nome"]);
      exit;
    }
    $users = read_json_file(USERS_FILE);
    if (!is_array($users)) $users = [];
    for ($i = 0; $i < count($users); $i++) {
      if (($users[$i]['group'] ?? '') === $oldName) {
        $users[$i]['group'] = $newName;
      }
    }
    persist_users_list($users);
    for ($i = 0; $i < count($groups); $i++) {
      if ($groups[$i] === $oldName) {
        $groups[$i] = $newName;
        break;
      }
    }
    $groups = normalize_groups($groups);
    persist_groups($groups);
    echo json_encode(["success" => true]);
    break;

  case 'DELETE':
    $name = trim((string)($payload['name'] ?? ''));
    if ($name === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Specificare il gruppo da eliminare"]);
      exit;
    }
    if (!in_array($name, $groups, true)) {
      http_response_code(404);
      echo json_encode(["success" => false, "message" => "Gruppo non trovato"]);
      exit;
    }
    $users = read_json_file(USERS_FILE);
    if (!is_array($users)) $users = [];
    foreach ($users as $user) {
      if (($user['group'] ?? '') === $name) {
        http_response_code(400);
        echo json_encode(["success" => false, "message" => "Impossibile eliminare un gruppo assegnato a degli utenti"]);
        exit;
      }
    }
    $index = array_search($name, $groups, true);
    if ($index !== false) {
      array_splice($groups, $index, 1);
    }
    persist_groups($groups);
    echo json_encode(["success" => true]);
    break;

  default:
    http_response_code(405);
    echo json_encode(["success" => false, "message" => "Metodo non supportato"]);
}

File: app/routers/groups.py
from __future__ import annotations
from typing import Any, Dict

from fastapi import APIRouter, Query
from psycopg.rows import dict_row

from ..common import success, error_response
from ..db import get_connection

router = APIRouter()

@router.get("/groups", response_model=None)
def list_groups(role: str = Query("user")):
    if role != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")
    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute("SELECT name FROM groups ORDER BY name")
            rows = cursor.fetchall()

    groups = [str(row["name"]) for row in rows]
    return success(groups=groups)

@router.post("/groups", response_model=None)
def create_group(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    name = str(payload.get("name", "")).strip()
    if not name:
        return error_response(400, "Il nome del gruppo è obbligatorio")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute("SELECT 1 FROM groups WHERE name = %s", (name,))
            if cursor.fetchone() is not None:
                return error_response(409, "Esiste già un gruppo con questo nome")

            cursor.execute("INSERT INTO groups (name) VALUES (%s)", (name,))
            conn.commit()

    return success()

@router.put("/groups", response_model=None)
def rename_group(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    old_name = str(payload.get("oldName", "")).strip()
    new_name = str(payload.get("newName", "")).strip()

    if not old_name or not new_name:
        return error_response(400, "Specificare i nomi del gruppo da modificare")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute("SELECT 1 FROM groups WHERE name = %s", (old_name,))
            if cursor.fetchone() is None:
                return error_response(404, "Gruppo non trovato")

            if old_name == new_name:
                return success()

            cursor.execute("SELECT 1 FROM groups WHERE name = %s", (new_name,))
            if cursor.fetchone() is not None:
                return error_response(409, "Esiste già un gruppo con il nuovo nome")

            cursor.execute(
                "UPDATE groups SET name = %s WHERE name = %s",
                (new_name, old_name),
            )
            conn.commit()

    return success()

@router.delete("/groups", response_model=None)
def delete_group(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    name = str(payload.get("name", "")).strip()
    if not name:
        return error_response(400, "Specificare il gruppo da eliminare")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute("SELECT 1 FROM groups WHERE name = %s", (name,))
            if cursor.fetchone() is None:
                return error_response(404, "Gruppo non trovato")

            cursor.execute(
                "SELECT COUNT(*) AS total FROM users WHERE group_name = %s",
                (name,),
            )
            row = cursor.fetchone()
            if row and row["total"] > 0:
                return error_response(
                    400, "Impossibile eliminare un gruppo assegnato a degli utenti"
                )

            cursor.execute("DELETE FROM groups WHERE name = %s", (name,))
            conn.commit()

    return success()